cpp/src/arrow/util/thread_pool.cc

```diff
@@ -44,6 +44,63 @@ struct Task {
 
 }  // namespace
 
+struct SerialExecutor::State {
+  std::queue<Task> task_queue;
+  std::mutex mutex;
+  std::condition_variable wait_for_tasks;
+};
+
+SerialExecutor::SerialExecutor() : state_(new State()) {}
+SerialExecutor::~SerialExecutor() {}
+
+Status SerialExecutor::SpawnReal(TaskHints hints, FnOnce<void()> task,
+                                 StopToken stop_token, StopCallback&& stop_callback) {
+  // The serial task queue is truly serial (no mutex needed) but SpawnReal may be called
+  // from external threads (e.g. when transferring back from blocking I/O threads) so a
+  // mutex is needed
+  {
+    std::lock_guard<std::mutex> lg(state_->mutex);
+    state_->task_queue.push(
+        Task{std::move(task), std::move(stop_token), std::move(stop_callback)});
+  }
+  state_->wait_for_tasks.notify_one();
+  return Status::OK();
+}
+
+void SerialExecutor::MarkFinished(bool& finished) {
+  {
+    std::lock_guard<std::mutex> lk(state_->mutex);
+    finished = true;
+  }
+  state_->wait_for_tasks.notify_one();
+}
+
+void SerialExecutor::RunLoop(const bool& finished) {
+  std::unique_lock<std::mutex> lk(state_->mutex);
+
+  while (!finished) {
+    while (!state_->task_queue.empty()) {
+      Task& task = state_->task_queue.front();
```

Comment:
The `task_queue` can be mutated from another thread, and I don't think there are any guarantees about reference stability.

Reply:
True, I thought I was fine since I re-lock before I pop but I suppose the underlying storage could be getting reallocated.

Reply:
Fixed.